README.md
# fumi

[![Clojars Project](https://img.shields.io/clojars/v/fumi.svg)](https://clojars.org/fumi)


A Prometheus client for Clojure.

Features:

- Simple wrapper of the official [java client](https://github.com/prometheus/client_java)
- Supports all prometheus metric types (counter, gauge, summary, histogram)
- Built in exporters for JVM / process metrics (additional dep required)
- Supports central configuration as well as creating collectors next to the code they instrument
- Supports custom collectors implemented as functions

## Status

fumi is used in production and the API should remain fairly stable. 
Please try it out and raise any issues you may find.

## Usage

For tools.deps, add this to your deps.edn:

```clojure
{fumi/fumi {:mvn/version "0.3.1"}}
```

For Leiningen, add this to your project.clj:

```clojure
[fumi "0.3.1"]
```

## Getting started

```clojure 
;; A default registry is already defined and exports jvm and process metrics. 
;; Use init! to redefine it, if you want to centralise configuration.
(init!
  {:default-exports?  true
   :collectors        {:test_counter   {:type :counter :help "a counter" :label-names [:foo]}
                       :test_gauge     {:type :gauge :help "a gauge"}
                       :test_histogram {:type :histogram :help "a histogram"}}})

;; Alternatively add collectors to the existing registry, useful
;; if you want to configure them next to the code they instrument
(register! :another_counter {:type :counter :help "another counter"})
(register! :test_summary {:type :summary :help "a summary"})

;; Observe some values
(increase! :test_counter {:n 3 :labels {:foo "bar"}})
(set-n! :test_gauge 2.1 {})
(observe! :test_histogram 0.51 {})

;; Output
(-> (collect)
    (serialize :text))
```

See the section on using self-managed registries for other usage patterns.

## Using other registries

Typically using the default registry is sufficient.
fumi also allows you to use separate registries or hold onto references. This may be useful for:

- injecting via component-style systems
- for using with pushgateway
- test isolation

In this case, the registry (i.e. result of calling `init!`) must be passed in as the `:registry` option
to all subsequent operations.

```clojure 
;; Create your registry by using the :self-managed? argument and holding on to the result
(def my-registry 
  (init! {:self-managed?     true
          :default-exports?  false
          :collectors        {:test_counter {:type :counter :help "a counter" :label-names [:foo]}}}))

;; Add more metrics to it
(register! :another_counter {:type :counter :help "another counter" :registry my-registry})

;; Observe some values
(increase! :test_counter {:n 3 :labels {:foo "bar"} :registry my-registry})

;; Output -- explicitly include default-registry if you want it
(-> (collect my-registry default-registry)
    (serialize :text))
```

## Metric types

fumi supports the standard prometheus [metric types](https://prometheus.io/docs/concepts/metric_types/)
(counter, gauge, summary, histogram). They are defined through a data driven API either during
initial setup of a registry via `init!`, or later on via `register!`.
 
In either case, the metric must be identified by a keyword (e.g. `:my_counter`), and takes these options:
- `:type` is a keyword corresponding to one of the metric types (`:counter`, `:gauge`, `:summary`, `:histogram`)
- `:help` is a short string description
- `:label-names` (optional) is a list of keywords. If supplied, any operation on the metric needs a 
  corresponding `:labels` map providing a value for each label-name

Histograms takes a further option:
- `:buckets` (optional), defaulting to `[0.005, 0.01, 0.025, 0.05, 0.075, 0.1, 0.25, 0.5, 0.75, 1, 2.5, 5, 7.5, 10]`

### Operations

These are documented in more details in the [API docs](https://cljdoc.org/d/fumi/fumi/CURRENT/api/fumi.client).

#### Counter

A counter can only increase.

```clojure
(increase! :my_counter)
(increase! :another_counter {:n 3 :labels {:uri "/home"}})
```

#### Gauge

A gauge can go up or down, or be set to a specific number.

```clojure
(increase! :my_gauge)
(decrease! :my_gauge {:n 2})
(set-n! :my_gauge 3.2)
```

#### Summary and Histogram

These metric types create a distribution based on observations. 

```clojure
(observe! :my_summary 3.2)
(observe! :my_histogram 4.7)
```

In all the above, if a metric was defined with one or more `:label-names`, an operation has to provide 
a `:labels` map with an entry for every label name.

## Generating output

Use `(collect)` to get the state of a registry at a point in time.

To satisfy the `/metrics` endpoint that a Prometheus server expects, create a route that returns
the data in text format:

```clojure
(-> (collect) 
    (serialize :text)) 

; Returns lines of text like:
; # HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
; # TYPE process_cpu_seconds_total counter
; process_cpu_seconds_total 18.976212
```

Note that `(collect)` with no arguments will collect from the default-registry, but if passed any arguments,
the default-registry needs to be explicitly provided:

```clojure
(collect my-registry default-registry)
```

## Built-in collectors

By default, fumi can expose [default collectors](https://github.com/prometheus/client_java#included-collectors) 
that expose several useful metrics:

- JVM: memory, threads, garbage collection, jvm info
- process: cpu, memory, file descriptors

To enable them, ensure you have the extra dependency:

```clojure
io.prometheus/simpleclient_hotspot {:mvn/version "0.12.0"}
```

`init!` the registry with `:default-exports?` set to true to add all of them.

To include them selectively, choose the relevant ones from those
[available](https://github.com/prometheus/client_java/blob/master/simpleclient_hotspot/src/main/java/io/prometheus/client/hotspot/DefaultExports.java#L37),
and provide them during `init!` or `register!`:

```clojure
; Add all default exports
(init! {:default-exports? true})

; Add only specific default exporters during init
(init! {:collectors {:version-info (io.prometheus.client.hotspot.VersionInfoExports.)}})

; Or to an existing registry
(register! :version-info (io.prometheus.client.hotspot.VersionInfoExports.))
```

## Custom collectors

On top of creating the standard metric collectors and performing operations on them based on some event,
it is possible to directly implement `Collector` abstract class, providing a function that returns the necessary data
as collection of metrics. You can directly pass in a function to `register!` or `init!` in place of an options map: 

```clojure
; Add a collector that a counter that is always stuck on 42.
(register! :constant_counter
           (fn [] [{:name    "metric"
                    :type    :counter
                    :help    "a constant counter"
                    :samples [{:value 42}]}
                   {:name    "metric_with_labels",
                    :type    :gauge,
                    :help    "a constant gauge with labels",
                    :samples [{:name   "active_requests",
                               :value  42
                               :labels {:source "foo"}}]}]))
```

You can also call `proxy-collector` with the same function to return a (not yet registered) `Collector` directly.

The provided function can return a single metric, or a collection of them.

## Usage with pushgateway

Pushgateway is useful for instrumenting short-lived processes (e.g. batch jobs) that may
not live long enough for metrics to be scraped.

To use it, simply send the serialized output of `collect` to the gateway using your http client of choice:

For example, using [hato](https://github.com/gnarroway/hato):
```clojure
(hato.client/post "http://your-gateway.com/metrics/job/somejob/somelabel/foo"
                  {:body (-> (fumi/collect)
                             (fumi/serialize :text))
                   :content-type "text/plain; version=0.0.4"})
```

## License

Released under the MIT License: http://www.opensource.org/licenses/mit-license.php


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [0.3.1]
### Fixed
- fix bucket creation for histogram (#4 thanks @aarongroom)

## [0.3.0]
### Changed
- **Breaking** Reimplemented internals as a wrapper of the Java client. This is to better prioritise performance
  as well as to keep up with changes to the monitoring ecosystem.
  The public API is mostly the same except as documented  below
- the need to opt *in* to `:default-exports?` (instead of opt *out* via `:exclude-defaults`)
- custom collectors can be created directly with a function without reifying a protocol
- the Java client automatically adds _created samples and _total suffixes (to counter samples) for open tracing compatibility
  These are passed through as they are so the final output may be slightly different.

## [0.2.0-b3] 2020-08-11
### Fixed
- Correct naming of process_max_fds (#1)

## [0.2.0-b2] 2020-04-14
### Changed
- More efficient implementations of Summary/Histogram

## [0.2.0-b1] 2020-04-08
### Changed
- Swapped implementation of counter/gauge for up to 10x faster DoubleAdder

## 0.1.0-beta1 - 2020-04-05
### Added
- Initial release

[Unreleased]: https://github.com/gnarroway/fumi/compare/v0.3.1...HEAD
[0.3.1]: https://github.com/gnarroway/fumi/compare/v0.3.0...v0.3.1
[0.3.0]: https://github.com/gnarroway/fumi/compare/v0.2.0-b3...v0.3.0
[0.2.0-b3]: https://github.com/gnarroway/fumi/compare/v0.2.0-b2...v0.2.0-b3
[0.2.0-b2]: https://github.com/gnarroway/fumi/compare/v0.2.0-b1...v0.2.0-b2
[0.2.0-b1]: https://github.com/gnarroway/fumi/compare/v0.1.0-beta1...v0.1.0-b1


project.clj
(defproject fumi "0.4.0-SNAPSHOT"
  :description "A Prometheus client for Clojure."
  :url "https://github.com/gnarroway/fumi"
  :license {:name         "The MIT License"
            :url          "http://opensource.org/licenses/mit-license.php"
            :distribution :repo}
  :deploy-repositories [["clojars" {:url           "https://clojars.org/repo"
                                    :username      :env/clojars_user
                                    :password      :env/clojars_pass
                                    :sign-releases false}]]
  :dependencies [[io.prometheus/simpleclient "0.12.0"]
                 [io.prometheus/simpleclient_common "0.12.0"]]
  :plugins [[lein-cljfmt "0.6.6"]]

  :profiles {:dev {:dependencies [[org.clojure/clojure "1.10.1"]
                                  [io.prometheus/simpleclient_hotspot "0.12.0"]]}})


src/fumi/collector/jvm.clj
(ns ^{:author "George Narroway"}
 fumi.collector.jvm
  "JVM collector"
  (:require [clojure.string :as string]
            [fumi.collector :as c])
  (:import (java.lang.management ManagementFactory ThreadInfo ThreadMXBean MemoryMXBean MemoryUsage GarbageCollectorMXBean)))

(set! *warn-on-reflection* true)

(def ^ThreadMXBean thread-bean (ManagementFactory/getThreadMXBean))
(def ^MemoryMXBean memory-bean (ManagementFactory/getMemoryMXBean))
(def gc-beans (ManagementFactory/getGarbageCollectorMXBeans))

(defn- memory-by-area
  [heap nonheap f]
  [{:value (f heap) :labels {:area "heap"}}
   {:value (f nonheap) :labels {:area "nonheap"}}])

(defn collect
  "Returns a store of JVM related metrics."
  []
  (let [collect-memory (partial memory-by-area
                                (.getHeapMemoryUsage memory-bean)
                                (.getNonHeapMemoryUsage memory-bean))]
    [;;; Memory
     {:name    :jvm_memory_bytes_init
      :type    :gauge
      :help    "Initial bytes of a given JVM memory area."
      :samples (collect-memory (fn [^MemoryUsage mu] (.getInit mu)))}

     {:name    :jvm_memory_bytes_used
      :type    :gauge
      :help    "Used bytes of a given JVM memory area."
      :samples (collect-memory (fn [^MemoryUsage mu] (.getUsed mu)))}

     {:name    :jvm_memory_bytes_committed
      :type    :gauge
      :help    "Committed bytes of a given JVM memory area."
      :samples (collect-memory (fn [^MemoryUsage mu] (.getCommitted mu)))}

     {:name    :jvm_memory_bytes_max
      :type    :gauge
      :help    "Max bytes of a given JVM memory area."
      :samples (collect-memory (fn [^MemoryUsage mu] (.getMax mu)))}

     ;;; Threads
     {:name    :jvm_threads_current
      :type    :gauge
      :help    "Current thread count of a JVM."
      :samples [{:value (.getThreadCount thread-bean)}]}

     {:name    :jvm_threads_daemon
      :type    :gauge
      :help    "Daemon thread count of a JVM."
      :samples [{:value (.getDaemonThreadCount thread-bean)}]}

     {:name    :jvm_threads_peak
      :type    :gauge
      :help    "Peak thread count of a JVM."
      :samples [{:value (.getPeakThreadCount thread-bean)}]}

     {:name    :jvm_threads_started_total
      :type    :gauge
      :help    "Started thread count of a JVM."
      :samples [{:value (.getTotalStartedThreadCount thread-bean)}]}

     {:name    :jvm_threads_deadlocked
      :type    :gauge
      :help    "Cycles of JVM threads that are in deadlock waiting to acquire object monitors or ownable synchronizers."
      :samples [{:value (count (.findDeadlockedThreads thread-bean))}]}

     {:name    :jvm_threads_deadlocked_monitor
      :type    :gauge
      :help    "Cycles of JVM threads that are in deadlock waiting to acquire object monitors."
      :samples [{:value (count (.findMonitorDeadlockedThreads thread-bean))}]}

     {:name    :jvm_threads_state
      :type    :gauge
      :help    "Current count of threads by state."
      :samples (->>
                (merge
                 (into {} (map (fn [^Thread$State t] [(.name t) 0]) (Thread$State/values)))
                 (frequencies
                  (map (fn [^ThreadInfo t] (-> t .getThreadState .name))
                       (.getThreadInfo ^ThreadMXBean thread-bean (.getAllThreadIds ^ThreadMXBean thread-bean) 0))))
                (map (fn [[k v]] {:value v :labels {:state k}})))}

     ;;; GC
     {:name    :jvm_gc_collection_seconds
      :type    :summary
      :help    "Time spent in a given JVM garbage collector in seconds."
      :samples (mapcat (fn [^GarbageCollectorMXBean gc]
                         [{:name   :jvm_gc_collection_seconds_count
                           :value  (.getCollectionCount gc)
                           :labels {:gc (.getName gc)}}
                          {:name   :jvm_gc_collection_seconds_sum
                           :value  (double (/ (.getCollectionTime gc) 1000))
                           :labels {:gc (.getName gc)}}])
                       gc-beans)}

     ;;; Other
     {:name    :jvm_info
      :type    :gauge
      :help    "JVM version info"
      :samples [{:value  1
                 :labels (->> ["java.runtime.version" "java.vm.vendor" "java.runtime.name"]
                              (map (juxt #(-> (string/split % #"\.") last) #(System/getProperty % "unknown")))
                              (into {}))}]}]))

(deftype JvmCollector []
  c/Collectable
  (-collect [_] (collect)))

(defn collector
  []
  (->JvmCollector))

src/fumi/collector/process.clj
(ns ^{:author "George Narroway"}
 fumi.collector.process
  "Process collector"
  (:require [clojure.string :as string]
            [fumi.collector :as c])
  (:import (java.lang.management ManagementFactory)))

(def ^:const milliseconds-per-second 1000.0)
(def ^:const nanoseconds-per-second 1e9)
(def ^:const kb 1024)

(def os-bean (ManagementFactory/getOperatingSystemMXBean))
(def runtime-bean (ManagementFactory/getRuntimeMXBean))

(defn- from-val
  [v]
  (when v [{:value v}]))

(defn- memory-stat
  "Extract stat from the content of /proc/self/status

  `stats` is a list of string tuples e.g. [[VMRSS:, 36, kb]]
  `field` is the content of the first tuple e.g. VMRSS:"
  [stats field]
  (some->> stats
           (filter #(= field (first %)))
           (first)
           (second)
           (Integer/parseInt)
           (* kb)))

(defn collect
  "Returns process related metrics."
  []
  (let [proc-status (try
                      (->> (slurp "/proc/self/status")
                           (string/split-lines)
                           (map #(string/split % #"\s+")))
                      (catch Throwable _
                        ; Expected for non-linux
                        nil))]
    [{:name    :process_cpu_seconds_total
      :type    :counter
      :help    "Total user and system CPU time spent in seconds."
      :samples (-> (.getProcessCpuTime os-bean)
                   (/ nanoseconds-per-second)
                   (from-val))}

     {:name    :process_start_time_seconds
      :type    :gauge
      :help    "Start time of the process since unix epoch in seconds."
      :samples (-> (.getStartTime runtime-bean)
                   (/ milliseconds-per-second)
                   (from-val))}

     {:name    :process_open_fds
      :type    :gauge
      :help    "Number of open file descriptors."
      :samples (-> (try
                     (.getOpenFileDescriptorCount os-bean)
                     (catch Throwable _
                       ; Expected for non-linux
                       nil))
                   (from-val))}

     {:name    :process_max_fds
      :type    :gauge
      :help    "Maximum number of open file descriptors."
      :samples (-> (try
                     (.getMaxFileDescriptorCount os-bean)
                     (catch Throwable _
                       ; Expected for non-linux
                       nil))
                   (from-val))}

     {:name    :process_resident_memory_bytes
      :type    :gauge
      :help    "Resident memory size in bytes."
      :samples (from-val (memory-stat proc-status "VmRSS:"))}

     {:name    :process_virtual_memory_bytes
      :type    :gauge
      :help    "Virtual memory size in bytes."
      :samples (from-val (memory-stat proc-status "VmSize:"))}]))

(deftype ProcessCollector []
  c/Collectable
  (-collect [_] (collect)))

(defn collector
  []
  (->ProcessCollector))

src/fumi/client.clj
(ns ^{:author "George Narroway"}
 fumi.client
  "Native clojure prometheus client"
  (:require [clojure.core :as core]
            [clojure.string :as string])
  (:import (io.prometheus.client Collector Collector$Type CollectorRegistry
                                 Collector$MetricFamilySamples Collector$MetricFamilySamples$Sample
                                 Counter Gauge Histogram Summary)
           (io.prometheus.client.exporter.common TextFormat)
           (java.io ByteArrayOutputStream)))

; We keep our own registry atom of the collectors so we can look them up by name
(defonce collectors (atom nil))
(defonce default-registry (CollectorRegistry/defaultRegistry))

; Enable registering default exports if the correct dependency is available
(defmacro include-optional-features []
  (if (try
        (import '(io.prometheus.client.hotspot DefaultExports))
        (catch ClassNotFoundException _ nil))
    '(defn- register-default-exports [registry]
       (DefaultExports/register registry))
    '(defn- register-default-exports [registry]
       (println "io.prometheus/simpleclient_hotspot must be added to use default exports"))))

(include-optional-features)

;;; Registry

(defn ->Type
  [type]
  (Collector$Type/valueOf (-> type
                              name
                              string/upper-case)))
(defn ->Sample
  [s]
  (let [ks (keys (:labels s))]
    (Collector$MetricFamilySamples$Sample. (some-> (:name s) name)
                                           ks
                                           (map (fn [k] (get (:labels s) k)) ks)
                                           (:value s))))

(defn ->MetricFamilySamples
  [x]
  (Collector$MetricFamilySamples. (some-> (:name x) name)
                                  (->Type (:type x))
                                  (:help x)
                                  (map ->Sample (:samples x))))

(defn proxy-collector
  "Turns a function return clojure datastructures into a Collector.

  The function should return a map (or list of maps) with fields:

  - `:name` the name of the metric or family
  - `:type` :counter, :gauge, :histogram, :summary. :info
  - `:help` a string describing the metric.
  - `:samples` a list of maps with keys:
    - `:name` (optional) the name of the metric to replace the parent name if provided.
    - `:value` the value of the metric
    - `:labels` (optional) a map of label-name to label value

  e.g.
  [{:name    \"jvm2\",
    :type    :info,
    :help    \"VM version info\",
    :samples [{:name   \"jvm_info\",
               :value  1.0,
               :labels {\"version\" \"11.0.11+9-LTS\"}}]}]"
  [f]
  (proxy
   [Collector] []
    (collect [& args]
      (let [res (f)]
        (if (map? res)
          (->MetricFamilySamples res)
          (map ->MetricFamilySamples res))))))

(defn ->Collector
  [name opts]
  (let [b (case (:type opts)
            :counter (Counter/build)
            :gauge (Gauge/build)
            :summary (Summary/build)
            :histogram (Histogram/build)
            (throw (IllegalArgumentException. "Unknown metric type")))
        ; name and help must always be set
        b (-> b
              (.name (core/name name))
              (.help (:help opts)))]

    (cond-> b
      (:label-names opts) (.labelNames (into-array String (map core/name (:label-names opts))))
      (and (= :histogram (:type opts))
           (:buckets opts)) (.buckets (double-array (map double (:buckets opts))))
      true (.create))))

(defn register!
  "Register a collector and returns a wrapped version of the collector with its options.

  Arguments:

  - `registry` a CollectorRegistry. if not provided, the default-registry will be used.
  - `name` is the name of the metric.
  - `opts` is a custom Collector, a function returning samples, or map of:
    - `:type` :counter, :gauge, :histogram, :summary.
    - `:help` a string describing the metric.
    - `:label-names` (optional) a list of string/keyword dimensions of this metric.
    - `:buckets` (optional, only for histogram), defaulting to `[0.005, 0.01, 0.025, 0.05, 0.075, 0.1, 0.25, 0.5, 0.75, 1, 2.5, 5, 7.5, 10]`.
    - `:registry` (optional) the result of calling init!. If not specified, uses the default-registry."
  ([name opts]
   (register! default-registry name opts))
  ([registry name opts]
   (let [registry (or registry default-registry)
         c (cond
             (map? opts) (->Collector name opts)
             (fn? opts) (proxy-collector opts)
             (instance? Collector opts) opts)]
     (.register registry c)
     (let [c' {:label-names (:label-names opts)
               :registry    registry
               :collector   c}]
       (swap! collectors assoc name c')
       c'))))

(defn init!
  "Initialises a registry with config and returns the registry.

  `config` is a map of:

  - `:self-managed?` (optional) if true, the default registry will not be used.
  - `:exclude-defaults?` (optional) exclude default collectors when set to true (default false).
  - `:collectors` a map of collector name to either a `Collectable`, or an options map with keys:
    - `:type` :counter, :gauge, :histogram, :summary.
    - `:help` a string describing the metric.
    - `:label-names` (optional) a list of string/keyword dimensions of this metric.
    - `:buckets` (optional, only for histogram), defaulting to `[0.005, 0.01, 0.025, 0.05, 0.075, 0.1, 0.25, 0.5, 0.75, 1, 2.5, 5, 7.5, 10]`."
  [config]
  (let [registry (if (:self-managed? config) (CollectorRegistry.) default-registry)]
    ; Clean up because the java client does not allow modification
    (.clear registry)
    (swap! collectors (fn [xs] (->> xs (remove (fn [[_ v]] (= registry (:registry v)))) (into {}))))
    (doseq [[name opts] (:collectors config)]
      (register! registry name opts))
    (when (:default-exports? config)
      (register-default-exports registry))
    registry))

;;; Operations

(defn- collector-or-child
  "Returns a collector to make observations on.
  Validates labels if the collector is configured with label-names.

  - `c` a name the collector was registered with, or a map of :collector :label-names"
  [c labels]
  (let [collector (if (or (keyword? c) (string? c)) (get @collectors c) c)]
    (when-not collector (throw (IllegalArgumentException. "collector does not exist")))
    (if-let [label-names (:label-names collector)]
      (do (when (not= (set label-names) (set (keys labels))) (throw (IllegalArgumentException. "labels must match label names")))
          (.labels (:collector collector) (into-array String (map (fn [k] (name (get labels k))) label-names))))
      (:collector collector))))

(defn increase!
  "Increase the value of a collector `name`.

  Arguments:

  - `c` a name the collector was registered with, or a collector itself
  - `opts` (optional), a map of:
    - `:n` a positive number to increase the value of the collector by (default 1)`\n
    - `:labels` a map of label name to value - must be provided if the collector has `label-names` defined."
  ([c] (increase! c {}))
  ([c opts]
   (let [{:keys [n labels] :or {n 1}} opts]
     (-> (collector-or-child c labels)
         (.inc n)))))

(defn decrease!
  "Decrease the value of collector `name`.

  Arguments:

  - `c` a name the collector was registered with, or a collector itself
  - `opts` (optional), a map of:
    - `:n` a positive number to decrease the value of the collector by (default 1)`
    - `:labels` a map of label name to value - must be provided if the collector has `label-names` defined."
  ([c] (decrease! c {}))
  ([c opts]
   (let [{:keys [n labels] :or {n 1}} opts]
     (-> (collector-or-child c labels)
         (.dec n)))))

(defn set-n!
  "Sets the value of a collector `name` to `n`.

  Arguments:

  - `c` a name the collector was registered with, or a collector itself
  - `n` a number to set the value of the collector to`
  - `opts` (optional), a map of:
    - `:labels` a map of label name to value - must be provided if the collector has `label-names` defined."
  ([c n] (set-n! c n {}))
  ([c n opts]
   (let [labels (:labels opts)]
     (-> (collector-or-child c labels)
         (.set n)))))

(defn observe!
  "Observe value `n` for collector `name`.

  Arguments:

  - `c` a name the collector was registered with, or a collector itself
  - `n` a number to set the value of the collector to`
  - `opts` (optional), a map of:
    - `:labels` a map of label name to value - must be provided if the collector has `label-names` defined."
  ([c n] (observe! c n {}))
  ([c n opts]
   (let [labels (:labels opts)]
     (-> (collector-or-child c labels)
         (.observe n)))))

(defn collect
  "Collects stats from one or more registries into a list of metrics.

  Uses the default registry if no arguments are provided.
  Otherwise, all desired registries (including the default) must be explicitly provided."
  ([] (collect default-registry))
  ([registry & registries]
   (->> (conj registries registry)
        (distinct)
        (mapcat (fn [registry] (enumeration-seq (.metricFamilySamples registry))))
        (map (fn [m] {:name    (.name m)
                      :type    (keyword (string/lower-case (.name (.type m))))
                      :help    (.help m)
                      :samples (->> (.samples m)
                                    (map (fn [s] (cond-> {:name  (.name s)
                                                          :value (.value s)}
                                                   (some? (.timestampMs s)) (assoc :ts (.timestampMs s))
                                                   (seq (.labelNames s)) (assoc :labels (zipmap (.labelNames s) (.labelValues s))))))
                                    (sort-by (juxt :name :value)))}))
        (sort-by :name))))

(defn- text-format-004
  [registry]
  (let [baos (ByteArrayOutputStream.)]
    (with-open [w (clojure.java.io/writer baos)]
      (TextFormat/write004 w (.metricFamilySamples registry)))
    (String. (.toByteArray baos))))

(defn collect-format
  "Collects stats from one or more registries into a list of metrics.
  Returns it as a string in the 0.0.4 text exposition format (to be extended in future)

  Uses the default registry if no arguments are provided.
  Otherwise, all desired registries (including the default) must be explicitly provided."
  ([] (collect-format nil))
  ([{:keys [registries] :or {registries [default-registry]}}]
   (->> registries
        (distinct)
        (map text-format-004)
        (string/join "\n"))))

;;; Serialization

(defn- label-pair
  [[k v]]
  (format "%s=\"%s\"" (name k) (name v)))

(defn- metric-row
  [parent-name {:keys [name labels value]}]
  (let [label-str (if (seq labels) (str "{" (string/join "," (map label-pair labels)) "}") "")]
    (format "%s%s %s" (core/name (or name parent-name)) label-str value)))

(defmulti serialize
  "Serialize collected metrics to output format."
  (fn [_ type] type))

(defmethod serialize :default
  [metrics _]
  (str (->> metrics
            (map #(string/join "\n" (concat [(format "# HELP %s %s" (name (:name %)) (:help %))
                                             (format "# TYPE %s %s" (name (:name %)) (name (if (= :info (:type %)) :gauge (:type %))))]

                                            (map (partial metric-row (:name %)) (:samples %)))))
            (string/join "\n\n"))
       "\n"))

(comment
  ; Use collectors directly
  (def c1 (->Collector "counter1" {:type :counter :help "help me"}))
  ; Need to register it before it gets exposed anywhere
  (.register c1)
  ; Make an observation
  (.inc c1)
  (increase! {:collector c1})

  ; Another one with labels
  (def c2 (->Collector "counter2" {:type :counter :help "help me" :label-names [:foo :bar]}))
  (.register c2)
  (-> c2
      (.labels (into-array String ["foo-v" "bar-v"]))
      (.inc 3))
  (increase! {:collector c2 :label-names [:foo :bar]} {:labels {:foo "1" :bar "2"}})

  ; Get the results
  (collect)
  (serialize *1 :text)

  ; Reset the registry
  (init! {})
  (init! {:collectors {:version-info (io.prometheus.client.hotspot.VersionInfoExports.)}})
  ; Add specific default exporters
  (register! :version-info (io.prometheus.client.hotspot.VersionInfoExports.))

  (register! :constant_counter
             (proxy-collector (fn [] [{:name    "metric"
                                       :type    :counter
                                       :help    "a constant counter"
                                       :samples [{:value 42}]}
                                      {:name    "metric_with_labels",
                                       :type    :gauge,
                                       :help    "a constant gauge with labels",
                                       :samples [{:name   "active_requests",
                                                  :value  42
                                                  :labels {:source "foo"}}]}])))

  ; Include all default exports
  (init! {:default-exports? true})

  (defn collector
    []
    [{:name    "metric",
      :type    :info,
      :help    "VM version info",
      :samples [{:value 1.0}]}
     {:name    "metric_with_sample_name",
      :type    :info,
      :help    "VM version info",
      :samples [{:name  "jvm_info",
                 :value 1.0}]}
     {:name    "metric_with_labels",
      :type    :counter,
      :help    "Some requests",
      :samples [{:name   "request_count",
                 :value  1.0
                 :labels {:source "foo"}}]}])

  (register! :foo (proxy-collector collector)))

(comment
  ;; Create the registry with defaults
  (init!
   {:default-exports? true
    :collectors       {:test_counter   {:type :counter :help "a counter" :label-names [:foo]}
                       :test_gauge     {:type :gauge :help "a gauge"}
                       :test_histogram {:type :histogram :help "a histogram"}}})

  ;; Add more metrics
  (register! :another_counter {:type :counter :help "another counter"})
  (register! :test_summary {:type :summary :help "a summary"})

  ;; Observe some values
  (increase! :test_counter {:n 3 :labels {:foo "bar"}})
  (increase! :test_counter {:n -1 :labels {:foo "bar" :moo "cow"}})
  (increase! :test_counter {:labels {:foo :bar}})

  (set-n! :test_gauge 2.1 {})
  (observe! :test_histogram 0.51 {})

  ;; Output
  (-> (collect)
      (serialize :text))
  ; Directly using java client output
  (collect-format))

(comment
  ;; Create separate registry
  (def my-registry (init!
                    {:self-managed?    true
                     :default-exports? false
                     :collectors       {:test_counter {:type :counter :help "a counter" :label-names [:foo]}}}))

  ;; Add more metrics
  (register! :another_counter {:type :counter :help "another counter" :registry my-registry})

  ;; Observe some values
  (increase! :test_counter {:n 3 :labels {:foo "bar"} :registry my-registry})

  ;; Output
  (-> (collect my-registry)
      (serialize :text)))
